Overhaul anonymous user system and first premium user creation

ATechAdventurer opened the pull request:
This PR fixes an issue where premium SSO users would be logged out automatically by the anonymous user. Created users also lacked the proper RBAC configuration and would be denied access to all parts of the builder app.

ATechAdventurer committed:
Refactor authentication setup to remove anonymous plugin and streamline Google OAuth handling

- Removed the anonymous plugin from auth-client and auth-config.
- Updated auth configuration to conditionally enable email/password authentication based on license type.
- Adjusted DataLoader to prevent automatic anonymous login for premium users with Google OAuth.
- Changed LoginButton to use email credentials for anonymous login.
- Cleaned up environment variables related to Google OAuth configuration.
- Enhanced plugin access logic to reflect new authentication strategy.

ATechAdventurer committed:
Enhance user management and permissions handling

- Added checks for the first premium user in both multi-user and single-user management plugins to grant system admin access.
- Updated user service to include methods for checking if a user is the first premium user and for granting system admin access.
- Improved ability creation logic to handle permissions more effectively based on user roles.
- Added error handling for missing permission rules in CASL resource clause.

ATechAdventurer committed:
Refactor user ability creation by removing redundant manage permission checks

- Eliminated the explicit management permissions for all resources when a user has 'manage' permission on 'all'.
- Streamlined the ability creation logic to focus on specific permissions provided in the input.

ATechAdventurer committed:
Refactor session typing in layout component

- Updated session typing in the getRootData function to use the imported Session type from auth/session.ts for improved type safety and clarity.

bears4barrett commented on app/lib/casl/user-ability.ts:
```diff
@@ -24,13 +24,47 @@ export type AppAbility = PureAbility<[PermissionAction, AppSubjects], PrismaQuer
+  if (hasManageAll) {
+    // Grant all actions on all resources
+    can('manage', 'all');
+    can('read', 'DataSource');
```
"manage" and "all" are special keywords in CASL, and should handle all of these permission grants without the need to individually set them.

bears4barrett commented on app/lib/casl/user-ability.ts:
```diff
@@ -24,13 +24,47 @@ export type AppAbility = PureAbility<[PermissionAction, AppSubjects], PrismaQuer
+
       case PermissionResource.BuilderApp:
+        can(action, resource);
+        break;
```
This is more verbose, but same logic as before.

```
case PermissionResource.all:
case PermissionResource.BuilderApp:
case PermissionResource.AdminApp:
  can(action, resource);
  break;
```

ATechAdventurer committed:
Refactor user management and session types for improved clarity

- Updated session types in auth/session.ts to be exported for better accessibility.
- Modified user management plugins to check for the first premium user without requiring a userId parameter.
- Enhanced userService to streamline the isFirstPremiumUser method by removing the userId argument.

bears4barrett commented on app/lib/services/userService.ts:
```diff
@@ -159,10 +185,135 @@ export const userService = {
+      // Grant full permissions to the role
+      const fullPermissions = [
+        // Manage all resources (for CASL compatibility)
+        { action: PermissionAction.manage, resource: PermissionResource.all },
```
This single permission grant handles all of the specific permissions, making them unnecessary.

ATechAdventurer committed:
Refactor user ability creation by removing unnecessary permission checks for specific resources

- Removed redundant permission checks for 'all' and 'BuilderApp' resources in the createAbilityForUser function.
- Streamlined the ability creation logic to focus on relevant permissions for improved clarity and efficiency.

ATechAdventurer committed:
Add centralized permission definitions and enhance user ability creation

- Introduced SYSTEM_ADMIN_PERMISSIONS to centralize permission definitions for system admin roles.
- Updated createAbilityForUser function to grant permissions based on centralized definitions, improving clarity and maintainability.
- Refactored userService to utilize centralized permissions for granting full access to system admin roles, reducing redundancy in permission checks.

ATechAdventurer committed:
Refactor user ability creation by removing centralized permission checks

- Removed the SYSTEM_ADMIN_PERMISSIONS logic from the createAbilityForUser function to simplify permission handling.
- Streamlined the ability creation process by focusing solely on the permissions provided in the input, enhancing clarity and maintainability.
